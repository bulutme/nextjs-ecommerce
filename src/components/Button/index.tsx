import styled, { css } from "styled-components";
import React, { ButtonHTMLAttributes } from "react";
import Spinner from "../Spinner";

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  label?: string;
  icon?: React.ReactNode;
  iconSize?: number;
  loading?: boolean;
  variant?: "default" | "outlined" | "dashed" | "link" | "filled" | "rounded";
  color?: "primary" | "error" | "warning" | "secondary" | "success" | "info";
  size?: "small" | "medium" | "large";
  fullWidth?: boolean;
}

const sizeStyles = {
  small: css`
    padding: 8px 16px;
    font-size: 14px;
  `,
  medium: css`
    padding: 12px 24px;
    font-size: 16px;
  `,
  large: css`
    padding: 16px 32px;
    font-size: 18px;
  `,
};

const colorStyles = {
  primary: css`
    background-color: ${({ theme }) => theme.colors.primary};
    color: ${({ theme }) => theme.colors.white};
  `,
  error: css`
    background-color: ${({ theme }) => theme.colors.error};
    color: ${({ theme }) => theme.colors.white};
  `,
  warning: css`
    background-color: ${({ theme }) => theme.colors.warning};
    color: ${({ theme }) => theme.colors.white};
  `,
  secondary: css`
    background-color: ${({ theme }) => theme.colors.secondary};
    color: ${({ theme }) => theme.colors.white};
  `,
  success: css`
    background-color: ${({ theme }) => theme.colors.success};
    color: ${({ theme }) => theme.colors.white};
  `,
  info: css`
    background-color: ${({ theme }) => theme.colors.info};
    color: ${({ theme }) => theme.colors.white};
  `,
};

const variantStyles = {
  default: css``,
  outlined: css`
    border: 2px solid ${({ theme }) => theme.colors.primary};
    background-color: transparent;
    color: ${({ theme }) => theme.colors.primary};
  `,
  dashed: css`
    border: 2px dashed ${({ theme }) => theme.colors.primary};
    background-color: transparent;
    color: ${({ theme }) => theme.colors.primary};
  `,
  link: css`
    background-color: transparent;
    color: ${({ theme }) => theme.colors.primary};
    text-decoration: underline;
  `,
  filled: css`
    background-color: ${({ theme }) => theme.colors.primary};
    color: #fff;
  `,
  rounded: css`
    border-radius: 4px;
  `,
};

const ButtonStyled = styled.button<ButtonProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  font-weight: ${({ theme }) => theme.fontWeight.medium};
  cursor: pointer;
  width: fit-content;
  position: relative;
  white-space: nowrap;
  transition: color 0.2s ease;
  border: none;
  outline: none;

  ${({ size = "medium" }) => sizeStyles[size]}
  ${({ color = "primary" }) => colorStyles[color]}
  ${({ variant = "default" }) => variantStyles[variant]}

  ${({ fullWidth }) =>
    fullWidth &&
    css`
      width: 100%;
    `}

  ${({ loading }) =>
    loading &&
    css`
      opacity: 0.5;
      pointer-events: none;
    `}

  ${({ disabled, loading }) =>
    (disabled || loading) &&
    css`
      opacity: 0.5;
      cursor: not-allowed;
    `}
`;

const Button: React.FC<ButtonProps> = ({
  icon,
  iconSize = 16,
  loading = false,
  children,
  ...rest
}) => {
  return (
    <ButtonStyled {...rest}>
      {icon && (
        <span
          style={{
            width: `${iconSize}px`,
            height: `${iconSize}px`,
            fontSize: `${iconSize ?? 24}px`,
            marginRight: "8px",
          }}
        >
          {icon}
        </span>
      )}
      {loading && <Spinner />}
      {children}
    </ButtonStyled>
  );
};

export default Button;
