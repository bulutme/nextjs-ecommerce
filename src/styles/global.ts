"use client";

import { createGlobalStyle } from "styled-components";

const GlobalStyles = createGlobalStyle`
  * {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
  }

  html {
    font-size: 16px;
  }

  html, body, #__next {
    height: 100%;
  }

  a {
    text-decoration: none;
    width: fit-content;
  }

  ul {
    list-style-type: none
  }
`;

export default GlobalStyles;
