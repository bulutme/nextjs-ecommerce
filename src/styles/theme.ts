export const theme = {
  colors: {
    black: "#000",
    white: "#fff",
    gray: "#808080",
    grayLight: "#d3d3d3",
    border: "#e2e8f0",
    primary: "#3bb77e",
    primaryDark: "#2b8a5e",
    error: "#dc2626",
    info: "#0891b2",
    success: "#16a34a",
  },
  screens: {
    "2xs": "400px",
    xs: "480px",
    sm: "576px",
    md: "768px",
    lg: "992px",
    xl: "1200px",
    "2xl": "1440px",
    "3xl": "1560px",
  },
  fontWeight: {
    normal: 400,
    medium: 500,
    semibold: 600,
    bold: 700,
  },
};
