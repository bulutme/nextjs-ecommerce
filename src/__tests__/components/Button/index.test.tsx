import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import Button, { ButtonProps } from "../../../components/Button";
import { theme } from "@/styles/theme";

const renderWithTheme = (
  component: React.ReactElement,
  options?: RenderOptions
) => render(<ThemeProvider theme={theme}>{component}</ThemeProvider>, options);

describe("Button Component", () => {
  const commonProps: ButtonProps = {
    color: "primary",
    size: "medium",
    $fullwidth: false,
  };

  test("renders correctly", () => {
    const { container } = renderWithTheme(
      <Button {...commonProps}>Click me</Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders correctly with icon and loading", () => {
    const { container } = renderWithTheme(
      <Button {...commonProps} icon={<span>ðŸš€</span>} loading>
        Click me
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders correctly with specific props", () => {
    const { container } = renderWithTheme(
      <Button {...commonProps} $variant="link" size="small" $fullwidth disabled>
        Click me
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
