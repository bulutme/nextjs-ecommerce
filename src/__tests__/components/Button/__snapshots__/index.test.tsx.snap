// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders correctly 1`] = `
<button
  class="sc-e6c76d24-0 gXeFVB"
  color="primary"
>
  Click me
</button>
`;

exports[`Button Component renders correctly with icon and loading 1`] = `
<button
  class="sc-e6c76d24-0 gXeFVB"
  color="primary"
>
  <span
    style="width: 16px; height: 16px; font-size: 16px; margin-right: 8px;"
  >
    <span>
      ðŸš€
    </span>
  </span>
  <div
    class="sc-ecd6f36e-0 eVitae"
  />
  Click me
</button>
`;

exports[`Button Component renders correctly with specific props 1`] = `
<button
  class="sc-e6c76d24-0 knoaAI"
  color="primary"
  disabled=""
>
  Click me
</button>
`;
